File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java
package fr.sncf.osrd.envelope_sim;

import static fr.sncf.osrd.envelope_sim.TrainPhysicsIntegrator.arePositionsEqual;
import static fr.sncf.osrd.envelope_utils.RangeMapUtils.fullyCovers;

import com.google.common.collect.ImmutableRangeMap;
import com.google.common.collect.Range;
import com.google.common.collect.RangeMap;
import com.google.common.collect.TreeRangeMap;
import fr.sncf.osrd.envelope_sim.electrification.Electrification;
import java.util.Arrays;
import java.util.Map;

public final class EnvelopeSimPath implements PhysicsPath {
    public final double length;

    /** The grade curve points */
    private final double[] gradePositions;

    /** The grade values between each pair of consecutive points */
    private final double[] gradeValues;

    /** The cumulative sum of the gradient at each grade position */
    private final double[] gradeCumSum;

    /**
     * A mapping describing electrification on this path (without electrical profiles nor
     * restrictions)
     */
    private final ImmutableRangeMap<Double, Electrification> defaultElectrificationMap;

    /**
     * Mapping from rolling stock power class to mapping describing electrification on this path
     * (without restrictions)
     */
    private final Map<String, ImmutableRangeMap<Double, Electrification>> electrificationMapByPowerClass;

    /**
     * Creates a new envelope path, which can be used to perform envelope simulations.
     *
     * @param length the length of the path
     * @param gradePositions the points at which the grade (slope) changes
     * @param gradeValues the values between consecutive pairs of grade positions
     * @param defaultElectrificationMap mapping from distance to electrification conditions
     * @param electrificationMapByPowerClass mapping from rolling stock power class to mapping from
     *     distance to electrification conditions
     */
    public EnvelopeSimPath(
            double length,
            double[] gradePositions,
            double[] gradeValues,
            ImmutableRangeMap<Double, Electrification> defaultElectrificationMap,
            Map<String, ImmutableRangeMap<Double, Electrification>> electrificationMapByPowerClass) {
        assert gradePositions.length == gradeValues.length + 1;
        assert gradePositions[0] == 0.0;
        assert gradePositions[gradePositions.length - 1] == length;
        for (int i = 0; i < gradePositions.length - 1; i++) assert gradePositions[i] < gradePositions[i + 1];
        this.gradePositions = gradePositions;
        this.gradeValues = gradeValues;
        this.length = length;
        this.gradeCumSum = initCumSum(gradePositions, gradeValues);
        assert fullyCovers(defaultElectrificationMap, length) : "default electrification map does not cover path";
        this.defaultElectrificationMap = defaultElectrificationMap;
        for (var entry : electrificationMapByPowerClass.entrySet())
            assert fullyCovers(entry.getValue(), length)
                    : "electrification map for power class " + entry.getKey() + " does not cover path";
        this.electrificationMapByPowerClass = electrificationMapByPowerClass;
    }

    private double[] initCumSum(double[] gradePositions, double[] gradeValues) {
        var result = new double[gradePositions.length];
        result[0] = 0.0;
        double cumSum = 0;
        for (int i = 0; i < gradePositions.length - 1; i++) {
            var rangeLength = gradePositions[i + 1] - gradePositions[i];
            cumSum += gradeValues[i] * rangeLength;
            result[i + 1] = cumSum;
        }
        return result;
    }

    @Override
    public double getLength() {
        return length;
    }

    private double getCumGrade(double position) {
        if (position > length && arePositionsEqual(position, length)) position = length;
        assert position <= length && position >= 0;
        var pointIndex = Arrays.binarySearch(gradePositions, position);
        if (pointIndex >= 0) return gradeCumSum[pointIndex];

        // when the position isn't found, binarySearch returns -(insertion point) - 1
        var insertionPoint = -(pointIndex + 1);

        // return the cumulative gradient at the point before the given position, plus the gradient
        // change since then
        var gradeRangeIndex = insertionPoint - 1;
        var gradeRangeStart = gradePositions[gradeRangeIndex];
        return gradeCumSum[gradeRangeIndex] + gradeValues[gradeRangeIndex] * (position - gradeRangeStart);
    }

    @Override
    public double getAverageGrade(double begin, double end) {
        if (begin == end) return getCumGrade(begin);
        return (getCumGrade(end) - getCumGrade(begin)) / (end - begin);
    }

    private RangeMap<Double, Electrification> getModeAndProfileMap(
            String powerClass, Range<Double> range, boolean ignoreElectricalProfiles) {
        if (ignoreElectricalProfiles) powerClass = null;
        return electrificationMapByPowerClass
                .getOrDefault(powerClass, defaultElectrificationMap)
                .subRangeMap(range);
    }

    /** Get the electrification related data for a given power class and power restriction map. */
    public ImmutableRangeMap<Double, Electrification> getElectrificationMap(
            String basePowerClass,
            RangeMap<Double, String> powerRestrictionMap,
            Map<String, String> powerRestrictionToPowerClass,
            boolean ignoreElectricalProfiles) {

        TreeRangeMap<Double, Electrification> res = TreeRangeMap.create();
        res.putAll(getModeAndProfileMap(basePowerClass, Range.closed(0.0, length), ignoreElectricalProfiles));

        if (powerRestrictionMap != null) {
            for (var entry : powerRestrictionMap.asMapOfRanges().entrySet()) {
                var restriction = entry.getValue();
                var powerClass = powerRestrictionToPowerClass.getOrDefault(restriction, basePowerClass);
                var modeAndProfileMap = getModeAndProfileMap(powerClass, entry.getKey(), ignoreElectricalProfiles);
                for (var modeAndProfileEntry : modeAndProfileMap.asMapOfRanges().entrySet()) {
                    var electrification = modeAndProfileEntry.getValue();
                    res.putCoalescing(modeAndProfileEntry.getKey(), electrification.withPowerRestriction(restriction));
                }
            }
        }
        return ImmutableRangeMap.copyOf(res);
    }

    /** Get the electrification related data for a given power class and power restriction map. */
    public ImmutableRangeMap<Double, Electrification> getElectrificationMap(
            String basePowerClass,
            RangeMap<Double, String> powerRestrictionMap,
            Map<String, String> powerRestrictionToPowerClass) {
        return getElectrificationMap(basePowerClass, powerRestrictionMap, powerRestrictionToPowerClass, false);
    }
}

File: core/kt-osrd-path/src/main/kotlin/fr/sncf/osrd/path/implementations/EnvelopeSimPath.kt
package fr.sncf.osrd.path.implementations

import fr.sncf.osrd.path.interfaces.Electrification
import fr.sncf.osrd.path.interfaces.PhysicsPath
import fr.sncf.osrd.utils.DistanceRangeMap
import fr.sncf.osrd.utils.OffsetRangeMap
import fr.sncf.osrd.utils.arePositionsEqual
import fr.sncf.osrd.utils.entries
import fr.sncf.osrd.utils.units.Distance
import fr.sncf.osrd.utils.units.meters
import java.util.Arrays

class EnvelopeSimPath(
    length: Double,
    gradePositions: DoubleArray,
    gradeValues: DoubleArray,
    defaultElectrificationMap: DistanceRangeMap<Electrification>,
    electrificationMapByPowerClass: Map<String, DistanceRangeMap<Electrification>>,
) : PhysicsPath {
    override val length: Double

    /** The grade curve points */
    private val gradePositions: DoubleArray

    /** The grade values between each pair of consecutive points */
    private val gradeValues: DoubleArray

    /** The cumulative sum of the gradient at each grade position */
    private val gradeCumSum: DoubleArray

    /**
     * A mapping describing electrification on this path (without electrical profiles nor
     * restrictions)
     */
    private val defaultElectrificationMap: DistanceRangeMap<Electrification>

    /**
     * Mapping from rolling stock power class to mapping describing electrification on this path
     * (without restrictions)
     */
    private val electrificationMapByPowerClass: Map<String, DistanceRangeMap<Electrification>>

    /**
     * Creates a new envelope path, which can be used to perform envelope simulations.
     *
     * @param length the length of the path
     * @param gradePositions the points at which the grade (slope) changes
     * @param gradeValues the values between consecutive pairs of grade positions
     * @param defaultElectrificationMap mapping from distance to electrification conditions
     * @param electrificationMapByPowerClass mapping from rolling stock power class to mapping from
     *   distance to electrification conditions
     */
    init {
        assert(gradePositions.size == gradeValues.size + 1)
        assert(gradePositions[0] == 0.0)
        assert(gradePositions[gradePositions.size - 1] == length)
        for (i in 0..<gradePositions.size - 1) assert(gradePositions[i] < gradePositions[i + 1])
        this.gradePositions = gradePositions
        this.gradeValues = gradeValues
        this.length = length
        this.gradeCumSum = initCumSum(gradePositions, gradeValues)
        assert(defaultElectrificationMap.fullyCovers(length.meters)) {
            "default electrification map does not cover path"
        }
        this.defaultElectrificationMap = defaultElectrificationMap
        for (entry in electrificationMapByPowerClass.entries) assert(
            entry.value.fullyCovers(length.meters)
        ) {
            "electrification map for power class " + entry.key + " does not cover path"
        }
        this.electrificationMapByPowerClass = electrificationMapByPowerClass
    }

    private fun initCumSum(gradePositions: DoubleArray, gradeValues: DoubleArray): DoubleArray {
        val result = DoubleArray(gradePositions.size)
        result[0] = 0.0
        var cumSum = 0.0
        for (i in 0..<gradePositions.size - 1) {
            val rangeLength = gradePositions[i + 1] - gradePositions[i]
            cumSum += gradeValues[i] * rangeLength
            result[i + 1] = cumSum
        }
        return result
    }

    private fun getCumGrade(position: Double): Double {
        var position = position
        if (position > length && arePositionsEqual(position, length)) position = length
        assert(position in 0.0..length)
        val pointIndex = Arrays.binarySearch(gradePositions, position)
        if (pointIndex >= 0) return gradeCumSum[pointIndex]

        // when the position isn't found, binarySearch returns -(insertion point) - 1
        val insertionPoint = -(pointIndex + 1)

        // return the cumulative gradient at the point before the given position, plus the gradient
        // change since then
        val gradeRangeIndex = insertionPoint - 1
        val gradeRangeStart = gradePositions[gradeRangeIndex]
        return gradeCumSum[gradeRangeIndex] +
            gradeValues[gradeRangeIndex] * (position - gradeRangeStart)
    }

    override fun getAverageGrade(begin: Double, end: Double): Double {
        if (begin == end) return getCumGrade(begin)
        return (getCumGrade(end) - getCumGrade(begin)) / (end - begin)
    }

    override fun getMinGrade(begin: Double, end: Double): Double {
        // TODO: Optimise method by adding in a cache.
        val indexBegin = getIndexBeforePos(begin)
        val indexEnd = getIndexBeforePos(end)
        // TODO: Remove if we extend path properties until last SvL > path.length.
        if (
            indexBegin == indexEnd && indexBegin == gradePositions.size - 1
        ) // Take last grade value in this case
         return gradeValues[gradeValues.size - 1]
        var lowestGradient = gradeValues[indexBegin]
        for (i in indexBegin..<indexEnd) {
            val grad = gradeValues[i]
            if (grad < lowestGradient) lowestGradient = grad
        }
        return lowestGradient
    }

    /** For a given position, return the index of the position just before in gradePositions */
    private fun getIndexBeforePos(position: Double): Int {
        // TODO: Optimise method by using binary search.
        if (position <= gradePositions[0]) return 0
        if (position >= gradePositions[gradePositions.size - 1]) return gradePositions.size - 1
        for (i in gradePositions.indices) {
            val pos = gradePositions[i]
            if (pos > position) return i - 1
        }
        return gradePositions.size - 1
    }

    private fun getModeAndProfileMap(
        powerClass: String?,
        lower: Distance,
        upper: Distance,
        ignoreElectricalProfiles: Boolean,
    ): DistanceRangeMap<Electrification> {
        val powerClass = if (ignoreElectricalProfiles) null else powerClass
        return electrificationMapByPowerClass
            .getOrDefault(powerClass, defaultElectrificationMap)
            .subMap(lower, upper)
    }

    /** Get the electrification related data for a given power class and power restriction map. */
    override fun getElectrificationMap(
        basePowerClass: String?,
        powerRestrictionMap: OffsetRangeMap<PhysicsPath, String>?,
        powerRestrictionToPowerClass: Map<String, String>?,
        ignoreElectricalProfiles: Boolean,
    ): DistanceRangeMap<Electrification> {
        val res =
            getModeAndProfileMap(basePowerClass, 0.meters, length.meters, ignoreElectricalProfiles)

        powerRestrictionMap?.forEach { lower, upper, restriction ->
            val powerClass = powerRestrictionToPowerClass?.getOrDefault(restriction, basePowerClass)
            val modeAndProfileMap =
                getModeAndProfileMap(
                    powerClass,
                    lower.distance,
                    upper.distance,
                    ignoreElectricalProfiles,
                )
            for (modeAndProfileEntry in modeAndProfileMap) {
                val electrification = modeAndProfileEntry.value
                res.put(
                    modeAndProfileEntry.lower,
                    modeAndProfileEntry.upper,
                    electrification.withPowerRestriction(restriction),
                )
            }
        }
        return res
    }
}
